module4/lessons/Week4/SecurityConcerns.md

```diff
@@ -0,0 +1,187 @@
+
+There are a number of [common hash function types](https://en.wikipedia.org/wiki/List_of_hash_functions) that are used in modern web development.  The one we will use today is SHA256.
+
+<section class='call-to-action' markdown='1'>
+
+Create a new console application and put this code in the `program.cs` file.  Run it a few times, and then annotate what each line of code. You will likely need to do some research in order to figure out what each line is doing!
+
+```c#
+HashAlgorithm sha = SHA256.Create();
+
+string firstInput = "Hello, World!";
+Console.WriteLine("Input: " + firstInput);
+byte[] firstInputBytes = Encoding.ASCII.GetBytes(firstInput);
+byte[] firstInputDigested = sha.ComputeHash(firstInputBytes);
+StringBuilder firstInputBuilder = new StringBuilder();
+foreach (byte b in firstInputDigested)
+{
+    Console.Write(b + ", ");
+    firstInputBuilder.Append(b.ToString("x2"));
+}
+Console.WriteLine();
+Console.WriteLine("Digest: " + firstInputBuilder);
+
+Console.WriteLine("\n\n");
+
+var secondInput = "Howdy";
+Console.WriteLine("Input: " + secondInput);
+byte[] secondInputBytes = Encoding.ASCII.GetBytes(secondInput);
+byte[] secondInputDigested = sha.ComputeHash(secondInputBytes);
+StringBuilder secondInputBuilder = new StringBuilder();
+foreach (byte b in secondInputDigested)
+{
+    Console.Write(b + ", ");
+    secondInputBuilder.Append(b.ToString("x2"));
+}
+Console.WriteLine();
+Console.WriteLine("Digest: " + secondInputBuilder);
+```
+
+Be ready to share your annotations!
+
+🌶️If you have time, are there ways that we could clean up this code?
+
+</section>
+
+Hashing data is a good way to securely store data, but it is not fool-proof.  Let's imagine hack into my application and are able to retrieve all my users’ hashed passwords. You find that the account with username `boss@example.com`` has this hashed password:
+
+```
+16e6b8af370f0d6b6e67c65bb6f4499c01a974f1dcd14a5f9429c6931531e6c4
+```
+
+Based on the length of the digest you guess it’s a SHA256. You know that some users, particularly bosses, are lazy and they do dumb things like re-use their 4-digit ATM pin for their password. But the application required a password of eight digits, so they might have repeated the pin.
+
+If these assumptions are correct, you are _very_ close to cracking into this user's account - it might only take you a few minutes of coding to find their password.
+
+To protect against this kind of bad data entry (people using easily guess-able password patterns).  Developers use a technique called **salting**.  A salt is an extra set of characters appended to any sensitive data before hashing.  It gives us longer, less predictable data. Check out [this blog post](https://auth0.com/blog/adding-salt-to-hashing-a-better-way-to-store-passwords/) for more info!
+
+
+
+### White Hat and Ethical Hacking
```

Comment:
I like that you have this section and also that it's just a small intro, I think that's all it needs!

I don't know if this analogy is helpful, but I like to think of it as an arms race between the software developers and the bad people trying to break in. One gets better, then the other gets better. And the white hat hackers are helping the developers!